firedrake/variational_solver.py

```diff
@@ -9,20 +9,25 @@
 from firedrake import utils
 from firedrake.petsc import PETSc, OptionsManager
 
-__all__ = ["LinearVariationalProblem",
+__all__ = ["VariationalProblem",
+           "LinearVariationalProblem",
            "LinearVariationalSolver",
            "NonlinearVariationalProblem",
            "NonlinearVariationalSolver"]
 
 
-class NonlinearVariationalProblem(object):
-    r"""Nonlinear variational problem F(u; v) = 0."""
+class VariationalProblem(object):
```

Comment:
Does it make sense to make a single `VariationalProblem` class, and then refactor the `NonlinearVariationalProblem` and `LinearVariationalProblem` into functional interfaces that just provide the correct arguments? (reducing the number of different classes).

Places where this might end up breaking things are in some of the multigrid coarsening/refining where we possibly do different things based on LVP vs NLVP.